src/edu/washington/escience/myria/operator/Limit.java

```diff
@@ -38,23 +41,30 @@ public Limit(@Nonnull final Long limit, final Operator child) {
     this.limit = Objects.requireNonNull(limit, "limit");
     Preconditions.checkArgument(limit >= 0L, "limit must be non-negative");
     toEmit = this.limit;
+    done = false;
   }
 
   @Override
   protected TupleBatch fetchNextReady() throws DbException {
     Operator child = getChild();
-    for (TupleBatch tb = child.nextReady(); tb != null; tb = child.nextReady()) {
-      if (tb.numTuples() <= toEmit) {
-        toEmit -= tb.numTuples();
-        return tb;
-      } else if (toEmit > 0) {
-        tb = tb.prefix(Ints.checkedCast(toEmit));
-        toEmit = 0;
-        return tb;
+    if (done) {
+      return null;
+    } else {
+      TupleBatch tb = child.nextReady();
+      TupleBatch result = null;
+      if (tb != null) {
+        if (tb.numTuples() <= toEmit) {
+          toEmit -= tb.numTuples();
+          result = tb;
+        } else if (toEmit > 0) {
```

Comment:
Actually I see that there's already a unit test as I described: `testLimitNumTuples`. It won't expose the bug though, since it doesn't test if the child has been closed after the limit is reached but before it is exhausted. You'll need something a little trickier to expose this bug--I would recommend using the `TupleSource(List<TupleBatch>)` constructor. You can generate two small `TupleBatch`es by creating two `TupleBatchBuffer`s with `TestUtils.range(limit)` and calling `popAny()` on each. Then initialize a `TupleSource` with a `List` of these two `TupleBatch`es and initialize a `Limit` operator with this `TupleSource`. Now call `nextReady()`, verify the `TupleBatch` returned has `numTuples() == limit`, and verify that the operator is at end-of-stream (`eos()`), which will be the case if it closed its child after the first call to `nextReady()`, as it should have.

@jingjingwang do you have a better idea for this test?